### PR TITLE
experiment: alpha/beta types in a single package with export paths

### PR DESCRIPTION
**Pull request by tylerbutler**
wip

## The scenario

We want to roll out some new APIs slowly over time. We intend to use the TSDoc `@alpha` and `@beta` tags to enable us to produce different types files that have some APIs trimmed out (this process is called "dts trimming").

However, there's only one `types` field in package.json, so we're stuck shipping different packages for different API types - e.g. an alpha-types package and a beta-types package, in addition to the standard package. While the user experience for customers isn't that bad, shipping different packages that differ only in superficial ways seems wasteful at best, and it also requires us to change packages in the CI pipeline far more than we do today. It would be far better if the package.json in the repo was almost exactly what we ship to npm.

## The concept

The idea is to use the `exports` field in package.json to output different paths that point to the same code, but different types. When importing from the package using a path like `import { API } from "@scope/package/beta";` the result would be the beta API, while the default path would omit those APIs.

## The exports field

Using this idea, the exports field for a CJS-and-ESM package would look something like this:

```json
"exports": {
  ".": {
    "import": {
      "types": "./lib/tree2-untrimmed.d.ts",
      "default": "./lib/index.js"
    },
    "require": {
      "types": "./dist/tree2-untrimmed.d.ts",
      "default": "./dist/index.js"
    }
  },
  "./alpha": {
    "import": {
      "types": "./lib/tree2-alpha.d.ts",
      "default": "./lib/index.js"
    },
    "require": {
      "types": "./dist/tree2-alpha.d.ts",
      "default": "./dist/index.js"
    }
  },
  "./beta": {
    "import": {
      "types": "./lib/tree2-beta.d.ts",
      "default": "./lib/index.js"
    },
    "require": {
      "types": "./dist/tree2-beta.d.ts",
      "default": "./dist/index.js"
    }
  }
},
```

### CODE REVIEW

File: examples/apps/tree-comparison/tsconfig.json [hunk 1]
```diff
@@ -1,7 +1,8 @@
 {
 	"extends": "@fluidframework/build-common/ts-common-config.json",
 	"compilerOptions": {
-		"module": "esnext",
+		"module": "node16",
+		"moduleResolution": "node16",
```
Comment (tylerbutler) [inline examples/apps/tree-comparison/tsconfig.json:5]: Without this the /alpha path wasn't found during compilation.

File: experimental/dds/tree2/package.json [hunk 2]
```diff
@@ -11,6 +11,38 @@
...
+	"exports": {
+		".": {
+			"import": {
+				"types": "./lib/tree2-untrimmed.d.ts",
+				"default": "./lib/index.js"
+			},
+			"require": {
+				"types": "./dist/tree2-untrimmed.d.ts",
+				"default": "./dist/index.js"
+			}
```
Comment (tylerbutler) [inline experimental/dds/tree2/package.json:23]: I used the untrimmed types here, and I wanted to use them in the top-level types field too. Unfortunately that breaks compilation of some packages, for reasons I don't yet understand. So I left the top-level one as-is. The exports values don't seem to be respected unless the typescript moduleResolution is node16, so no other packages see the untrimmed types yet.

File: fluidBuild.config.cjs [hunk 3]
```diff
@@ -3,7 +3,7 @@
  * Licensed under the MIT License.
  */
 
-const tscDependsOn = ["^tsc", "build:genver"];
+const tscDependsOn = ["^tsc", "^api-extractor:commonjs", "^api-extractor:esnext", "build:genver"];
```
Comment (tylerbutler) [inline fluidBuild.config.cjs:6]: Builds now require api-etractor to run beforehand and create the rollup types.

**Comment by sonalideshpandemsft** [comment]
Once this change is implemented, we won't need the setPackageTypesField flub command, right?

**Comment by tylerbutler** [comment]
Superseded by other PRs.

**Pull request closed without merging**